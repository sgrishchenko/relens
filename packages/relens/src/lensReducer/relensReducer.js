import { set } from '../helpers/helpers';
import { RELENS_UPDATE } from '../lensAction/lensAction';

const relensReducer = reducer => (state, action) => {
  if (action.type === RELENS_UPDATE) {
    const { path, value } = action.payload;
    return set(path, value, state);
  }
  if (typeof reducer === 'function') {
    return reducer(state, action);
  }

  return state;
};

export default relensReducer;
