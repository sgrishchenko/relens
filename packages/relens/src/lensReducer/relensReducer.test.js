import update, { RELENS_UPDATE } from '../lensAction/lensAction';
import relensReducer from './relensReducer';

describe('relensReducer update', () => {
  test(`should accept the action with the type "${RELENS_UPDATE}" and set the value along the specified path`, () => {
    const value = 'value';
    const action = update(['some', 'deep', 'prop'])(value);
    const actualState = relensReducer()(undefined, action);

    expect(actualState).toMatchSnapshot();
  });

  test('should proxy the internal reducer if it is specified', () => {
    const EXAMPLE_ACTION_TYPE = 'EXAMPLE_ACTION_TYPE';
    const action = {
      type: EXAMPLE_ACTION_TYPE,
      payload: 'some payload',
    };
    const initialState = { prop: 'some initial value' };
    const resultState = { prop: 'some state value' };
    const internalReducer = jest.fn(() => resultState);
    const actualState = relensReducer(internalReducer)(initialState, action);

    expect(actualState).toBe(resultState);
    expect(internalReducer).toHaveBeenCalledWith(initialState, action);
  });

  test('should not affect the values of other fields of the previous state', () => {
    const value = 'value';
    const action = update(['some', 'deep', 'prop'])(value);
    const previousState = { prop: 'some other value' };
    const actualState = relensReducer()(previousState, action);

    expect(actualState).toEqual(
      expect.objectContaining(previousState),
    );
    expect(actualState).toMatchSnapshot();
  });

  test('should return the previous state otherwise', () => {
    const EXAMPLE_ACTION_TYPE = 'EXAMPLE_ACTION_TYPE';
    const action = {
      type: EXAMPLE_ACTION_TYPE,
      payload: 'some payload',
    };
    const previousState = { prop: 'some other value' };
    const actualState = relensReducer()(previousState, action);

    expect(actualState).toEqual(previousState);
  });
});
