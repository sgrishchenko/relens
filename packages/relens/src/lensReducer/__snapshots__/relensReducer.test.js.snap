// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relensReducer update should accept the action with the type "RELENS_UPDATE" and set the value along the specified path 1`] = `
Object {
  "some": Object {
    "deep": Object {
      "prop": "value",
    },
  },
}
`;

exports[`relensReducer update should not affect the values of other fields of the previous state 1`] = `
Object {
  "prop": "some other value",
  "some": Object {
    "deep": Object {
      "prop": "value",
    },
  },
}
`;
